src/docs/arch-block.dox

```diff
@@ -1,9 +1,132 @@
+be interleaved).
+
+@section block_usage How WiredTiger uses the block manager
+
+Whenever a page is read into the cache, WiredTiger will read its corresponding
+block from the data file. This block contains what's referred to as the
+"page image" in other subsystems and has the newest versions of data for each
+key.
+
+Conversely, when a page is reconciled whether that is due to a page eviction or
+by a checkpoint, a new page image is constructed and it is written as a new
+block in the data file. In the case of checkpoint, the newly written block is
+bound to the current checkpoint and in the case of eviction, it is associated
+with the next checkpoint.
+
+@section block_implementation How the block manager is implemented
+
+We've already discussed the concept of a block and how the data file's on-disk
+representation is divided into smaller blocks that are each associated with a
+checkpoint.
+
+In this section, we'll dive into the file layout in more detail and talk about
+some of the core data structures used as part of block allocation.
+
+@subsection block_layout Block layout
+
+Each block consists of a page header, followed by a block manager header,
+followed by the actual page image.
+
+The page header contains some metadata about the page contents such as the type
+of page, whether it is compressed, whether it is encrypted, its write generation
+, etc.
+
+The block manager header is simpler. It contains the size of the following disk
+image as well as an optional checksum for when block-level checksums are
+enabled.
+
+@subsection desc_block Descriptor block
+
+As discussed earlier, a file is divided up into blocks. The first block in a file
+is special as it contains metadata about the file and is referred to as the
+"descriptor block". It contains the WiredTiger major and minor version, a
+checksum of the block contents as well as a "magic" number to check against.
+
+The descriptor block serves as a safety check to ensure that the file being
+loaded into the block manager is actually a WiredTiger data file, that it
+belongs to a compatible version of WiredTiger and that the entire file has not
+been corrupted. WiredTiger also offers block-level checksums to defend against
+file corruption at a more granular level, which can be found in the
+\ref checksum section.
+
+@subsection block_lists Extent lists
+
+Internally, WiredTiger uses a data structure called an extent list or a
+\c WT_EXTLIST to track file usage. An extent list consists of a series of
+extents (or \c WT_EXT elements). Each extent tracks a file offset and size to
+track a portion of the file.
+
+There are three extent lists that are maintained per checkpoint:
+
+- \c alloc: The extents allocated for a given checkpoint.
+- \c avail: The extents that are unused and available for allocation.
+- \c discard: The extents freed in the current checkpoint. After the checkpoint
+completes, these extents are shifted to the \c avail extent list.
```

Comment:
> Keith B: Because blocks can appear in multiple checkpoints, you have to expect an allocation in one checkpoint, then lots of intermediate checkpoints using it, and then an eventual discard in another checkpoint.
>
> Imagine we allocate a block in checkpoint 37, do nothing at all with that block for checkpoints 38-41, then delete the block in checkpoint 42. The block will appear in checkpoint 37's alloc list, won't appear in any list in checkpoints 38-41, and then appears in checkpoint 42's discard list.
> 
> As each checkpoint is deleted, the blocks it allocated move into the next checkpoint's list (so, when I delete checkpoint 37, the blocks in its alloc list are merged into the next checkpoint's alloc list).
> 
> When a block appears on both the alloc & discard lists in a single checkpoint, it moves to the avail list, because that means no checkpoint is still using it.
> 
> In other words, once a block is allocated, it cannot be reused until it's deleted, and no intermediate checkpoints are still using it.